### PR TITLE
dev/core#2721 Move custom search framework to a core extension

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
dev/core#2721 Move custom search framework to a core extension

Before
----------------------------------------
Code to support custom searches hacked into core search controller & state machine classes

After
----------------------------------------
Separated out into a core extension, hacks removed

Technical Details
----------------------------------------
In review for some of Matt's recent fixes for the core controller I took a look at https://github.com/civicrm/civicrm-core/pull/19143 & realised that it was adding hacks where we should be removing them and that doing the separation would unblock that. This extensionisation must remain hidden for now because

1) extensions need time to add this as a 'required' for their extensions
2) we need to figure out a hook or interaction to allow us to remove `getCustomSearchSQL` from `CRM_Contact_BAO_GroupContactCache`
3) we need to decide what to do about the custom searches that are added on install - ie some of them are entirely redundant as a result of search kit
4) I suspect some more stuff will still need to go over

Comments
----------------------------------------

https://lab.civicrm.org/dev/core/-/issues/2721

### CODE REVIEW

File: xml/templates/civicrm_data.tpl [hunk 1]
```diff
@@ -1791,4 +1791,5 @@ INSERT IGNORE INTO civicrm_extension (type, full_name, name, label, file, is_act
 INSERT IGNORE INTO civicrm_extension (type, full_name, name, label, file, is_active) VALUES ('module', 'contributioncancelactions', 'Contribution cancel actions', 'Contribution cancel actions', 'contributioncancelactions', 1);
 INSERT IGNORE INTO civicrm_extension (type, full_name, name, label, file, is_active) VALUES ('module', 'recaptcha', 'reCAPTCHA', 'reCAPTCHA', 'recaptcha', 1);
 INSERT IGNORE INTO civicrm_extension (type, full_name, name, label, file, is_active) VALUES ('module', 'ckeditor4', 'CKEditor4', 'CKEditor4', 'ckeditor4', 1);
+INSERT IGNORE INTO civicrm_extension (type, full_name, name, label, file, is_active) VALUES ('module', 'ckeditor4', 'Custom Search framework', 'Custom Search framework', 'legacycustomsearches', 1);
```
Comment (seamuslee001) [inline xml/templates/civicrm_data.tpl:1794]: @eileenmcnaughton I think you have the fullname wrong here?

File: CRM/Upgrade/Incremental/php/FiveFortyOne.php [hunk 2]
```diff
@@ -60,16 +60,39 @@ public function setPostUpgradeMessage(&$postUpgradeMessage, $rev) {
...
   //   * @param string $rev
   //   */
-  //  public function upgrade_5_0_x($rev) {
-  //    $this->addTask(ts('Upgrade DB to %1: SQL', [1 => $rev]), 'runSql', $rev);
-  //    $this->addTask('Do the foo change', 'taskFoo', ...);
-  //    // Additional tasks here...
-  //    // Note: do not use ts() in the addTask description because it adds unnecessary strings to transifex.
-  //    // The above is an exception because 'Upgrade DB to %1: SQL' is generic & reusable.
-  //  }
+  public function upgrade_5_40_alpha1($rev) {
```
Comment (seamuslee001) [inline CRM/Upgrade/Incremental/php/FiveFortyOne.php:63]: Function name needs to be fixed

File: xml/templates/civicrm_data.tpl (hunk 1 above)
Comment (eileenmcnaughton) [reply xml/templates/civicrm_data.tpl]: thanks - I think I got it

File: CRM/Upgrade/Incremental/php/FiveFortyOne.php (hunk 2 above)
Comment (eileenmcnaughton) [reply CRM/Upgrade/Incremental/php/FiveFortyOne.php]: fixed

**Comment by colemanw** [comment]
Code looks good to me

**Comment by demeritcowboy** [comment]
It crashes on the test site here if you try to run any custom search.

```
Exception: "Cannot determine api action for CRM_Contact_Form_Search_Basic.CRM_Core_Action "copy" not recognized."

 0 /home/jenkins/bknix-dfl/build/core-20952-5v7z7/web/sites/all/modules/civicrm/CRM/Core/Form.php(1622): CRM_Core_Form->getApiAction()
 1 /home/jenkins/bknix-dfl/build/core-20952-5v7z7/web/sites/all/modules/civicrm/CRM/Contact/Form/Search/Basic.php(58): CRM_Core_Form->addField("group", (Array:6))
 2 /home/jenkins/bknix-dfl/build/core-20952-5v7z7/web/sites/all/modules/civicrm/CRM/Core/Form.php(642): CRM_Contact_Form_Search_Basic->buildQuickForm()
 3 /home/jenkins/bknix-dfl/build/core-20952-5v7z7/web/sites/all/modules/civicrm/CRM/Core/QuickForm/Action/Display.php(76): CRM_Core_Form->buildForm()
 4 /home/jenkins/bknix-dfl/build/core-20952-5v7z7/web/sites/all/modules/civicrm/packages/HTML/QuickForm/Controller.php(203): CRM_Core_QuickForm_Action_Display->perform(Object(CRM_Contact_Form_Search_Basic), "display")
 5 /home/jenkins/bknix-dfl/build/core-20952-5v7z7/web/sites/all/modules/civicrm/packages/HTML/QuickForm/Page.php(103): HTML_QuickForm_Controller->handle(Object(CRM_Contact_Form_Search_Basic), "display")
 6 /home/jenkins/bknix-dfl/build/core-20952-5v7z7/web/sites/all/modules/civicrm/CRM/Core/Controller.php(352): HTML_QuickForm_Page->handle("display")
 7 /home/jenkins/bknix-dfl/build/core-20952-5v7z7/web/sites/all/modules/civicrm/CRM/Core/Invoke.php(313): CRM_Core_Controller->run((Array:4), (Array:0))
 8 /home/jenkins/bknix-dfl/build/core-20952-5v7z7/web/sites/all/modules/civicrm/CRM/Core/Invoke.php(69): CRM_Core_Invoke::runItem((Array:12))
 9 /home/jenkins/bknix-dfl/build/core-20952-5v7z7/web/sites/all/modules/civicrm/CRM/Core/Invoke.php(36): CRM_Core_Invoke::_invoke((Array:4))
 10 /home/jenkins/bknix-dfl/build/core-20952-5v7z7/web/sites/all/modules/civicrm/drupal/civicrm.module(458): CRM_Core_Invoke::invoke((Array:4))
 11 /home/jenkins/bknix-dfl/build/core-20952-5v7z7/web/includes/menu.inc(527): civicrm_invoke("contact", "search", "custom")
 12 /home/jenkins/bknix-dfl/build/core-20952-5v7z7/web/index.php(21): menu_execute_active_handler()
 13 {main}
```

**Comment by eileenmcnaughton** [comment]
@demeritcowboy hmm - looks like my follow on cleanup was too aggressive - I put back the lines that I think went too far

**Comment by eileenmcnaughton** [comment]
test this please

(unrelated fail)

**Comment by demeritcowboy** [comment]
Warning: require_once(CRM/Contact/Form/Search/Custom.php): failed to open stream: No such file or directory in require_once() (line 452 of /home/jenkins/bknix-dfl/build/core-20952-5vmuh/web/sites/all/modules/civicrm/CRM/Core/Controller.php).

**Comment by eileenmcnaughton** [comment]
Hmm - I'm not getting that - but when I do
```
SELECT page_callback FROM civicrm_menu WHERE path =
'civicrm/contact/search/custom';
```

I get

image


That clearly isn't the case on the demo site

**Comment by eileenmcnaughton** [comment]
I've updated civicrm_generated.mysql - let's see if that helps

**Comment by seamuslee001** [comment]
@eileenmcnaughton it is looking healthier on the demo site now but not sure if related or not

```
User warning: Smarty error: unable to read resource: "CRM/Contact/Form/Search/Custom/ActivitySearch.hlp" in Smarty->trigger_error() (line 1100 of /home/jenkins/bknix-dfl/build/core-20952-5w7ic/web/sites/all/modules/civicrm/packages/Smarty/Smarty.class.php).
```

**Comment by eileenmcnaughton** [comment]
yeah - that will be! I feel like those help files might dish up some challenges

**Comment by demeritcowboy** [comment]
I don't think the help file itself is the problem it's that the search results screen is including a link to advanced search whereas it doesn't before the patch.

**Comment by eileenmcnaughton** [comment]
it seems action needs to be 16384 (COPY) to render correctly

**Comment by eileenmcnaughton** [comment]
OK - I think I found it

**Comment by demeritcowboy** [comment]
This looks better just wondering about how load order works, which I'll try testing. For example if another extension implements a custom search, does it fail if it loads first before this extension.

**Comment by eileenmcnaughton** [comment]
@demeritcowboy I think that doesn't matter in that
1) for the next 6+ months this should be an always installed & hidden extension
2) that time should allow other extensions to add the <requires> tag to their html

**Comment by demeritcowboy** [comment]
Ok. I meant something more like extension X has a search which extends CRM_Contact_Form_Search_Custom_Base which used to be in core but is now in an extension, but extension X gets "loaded" or whatever the word is before extension legacycustomsearches gets loaded. But I'm probably just unclear on how that ordering works. I did test anyway because that's what I do™, and nothing obvious came up.

So I don't have anything else.

**Comment by eileenmcnaughton** [comment]
I guess loading in this cases means adding the extension to the path - which probably happens before any attempt to load a specific form - although I don't know all the details

**Comment by eileenmcnaughton** [comment]
squished

**Comment by seamuslee001** [comment]
@eileenmcnaughton are you able to rebase this?

**Comment by eileenmcnaughton** [comment]
Done - tests running again

**Comment by eileenmcnaughton** [comment]
cool -  I might re-try some of that cleanup since I suspect the last fix was the reason it went wonky

**Pull request merged**